db/migrate/20231210061330_add_created_at_to_fts_targets_with_index.rb

```diff
@@ -0,0 +1,105 @@
+      update_created_at("WikiPage", "wiki_pages", "created_on")
+    end
+
+    if Redmine::Database.mysql?
+      add_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :created_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+  end
+
+  def down
+    if Redmine::Database.mysql?
+      remove_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+
+    remove_column :fts_targets, :created_at
+  end
+
+  private
+
+  def update_created_at_for_changes
+    execute <<-SQL
+      UPDATE fts_targets
+      SET created_at = (SELECT committed_on FROM changesets WHERE changesets.id = (SELECT changeset_id FROM changes WHERE changes.id = fts_targets.source_id))
+      WHERE fts_targets.source_type_id = (SELECT id FROM fts_types WHERE name = 'Change')
+    SQL
+  end
+
+  def update_created_at_for_custom_values
+    execute <<-SQL
+      UPDATE fts_targets
+      SET created_at = (SELECT created_on FROM issues WHERE issues.id = (SELECT customized_id FROM custom_values WHERE custom_values.id = fts_targets.source_id AND customized_type = 'Issue'))
+      WHERE fts_targets.source_type_id = (SELECT id FROM fts_types WHERE name = 'CustomValue') AND
+            EXISTS (SELECT 1 FROM custom_values WHERE custom_values.id = fts_targets.source_id AND customized_type = 'Issue')
+    SQL
+
+    execute <<-SQL
+      UPDATE fts_targets
+      SET created_at = (SELECT created_on FROM projects WHERE projects.id = (SELECT customized_id FROM custom_values WHERE custom_values.id = fts_targets.source_id AND customized_type = 'Project'))
+      WHERE fts_targets.source_type_id = (SELECT id FROM fts_types WHERE name = 'CustomValue') AND
+            EXISTS (SELECT 1 FROM custom_values WHERE custom_values.id = fts_targets.source_id AND customized_type = 'Project')
+    SQL
```

Comment:
Can we simplify this and other `UPDATE`?

```sql
UPDATE fts_targets
SET created_at = last_modified_at
WHERE fts_targets.source_type_id IN (SELECT id FROM fts_types WHERE name IN ('CustomValue', ...));
```

BTW, can we use Active Record API instead of direct SQL?

Reply:
> Can we use Active Record API instead of direct SQL?

I don't use it because it will raise errors if we delete a related Active Record model in the future.
Maybe I misunderstood what you meant. If so, please let me know.
- ref: https://blog.appsignal.com/2020/11/18/rails-model-patterns-and-anti-patterns.html#try-to-avoid-active-record-in-migrations

Reply:
Or if we define the model like the following in this migration file, I think we can use the Active Record API.
```ruby
class TempTarget < ActiveRecord::Base
    self.table_name = 'ftg_targets'
end
```

Reply:
I thought the migration file local Active Record model definition approach.
I want to avoid embedding variable to SQL without escaping like https://github.com/clear-code/redmine_full_text_search/pull/118/files#diff-b79e9e1bd296b63bcb985d654a3f1b308e718648a67719b893c0da9f03627fa5R101-R102 for safety.

Reply:
fix: eab63c4...c20eb78
I have concerns about the number of SQL queries generated by the following method, as it will produce the same number of queries as there are target records. Could you have a good idea?
- `update_created_at_using_created_on`
- `update_created_at_for_changes`

Reply:
fix: [5508d4c](https://github.com/clear-code/redmine_full_text_search/pull/118/commits/5508d4c3cdc049ac687738cc2889e3b39ab88aa5) I tried to reduce the number of generated queries but it still has a lot mm